rpc: don't leave poison zero-nodeID connections in pool

tbg opened the pull request:
An optimiziation to share the `(target,remoteNodeID)` connection under a
second name `(target,0)` backfired because we were never unregistering
the latter, meaning that clients requesting `(target,0)` would be handed
an eternally broken connection.

See #37200.

Release note (bug fix): Avoid a source of internal connectivity problems
that would resolve after restarting the affected node.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/37204)
<!-- Reviewable:end -->

tbg committed:
rpc: don't leave poison zero-nodeID connections in pool

An optimiziation to share the `(target,remoteNodeID)` connection under a
second name `(target,0)` backfired because we were never unregistering
the latter, meaning that clients requesting `(target,0)` would be handed
an eternally broken connection.

See #37200.

Release note (bug fix): Avoid a source of internal connectivity problems
that would resolve after restarting the affected node.

knz reviewed (approved):
Nice find and elegant fix. I suppose this was precisely the kind of problem Peter was foreseeing when I initially made the change. I'm sorry I did not consider this before. LGTM in any case. Thank you!

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37204#-:-Ldi7ESx8hTdtMQ-kMir:b-cved07)* status: :shipit: complete! 0 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

tbg commented:
> I'm sorry I did not consider this before

Well, I'm sorry I didn't see it in review :-) It happens.

bors r=knz
